src/main/java/org/reactome/addlinks/db/ReferenceObjectCache.java

```diff
@@ -106,183 +106,217 @@ private static synchronized void buildReferenceCaches(String className, Map<Stri
-					}
-					
-				}
-				if (cacheBySpecies != null)
-				{
-					for (String species : allSpecies)
-					{
-						//Add to the Species Cache
-						if (! cacheBySpecies.containsKey(species))
-						{
-							List<GKInstance> bySpecies = Collections.synchronizedList(new LinkedList<GKInstance>());
-							bySpecies.add(referenceObject);
-							cacheBySpecies.put(species, bySpecies);
-						}
-						else
-						{
-							cacheBySpecies.get(species).add(referenceObject);
-						}
-					}
-				}
+				addObjectToSpeciesCache(className, cacheBySpecies, referenceObject);
 			}
 			//ReferenceDatabase Cache
 			//
 			if (cacheByRefDB != null)
 			{
-				String refDBID;
+				addObjectToRefDBCache(cacheByRefDB, referenceObject);
+			}
+			// DB_ID Cache
+			cacheByID.put(String.valueOf(referenceObject.getDBID()), referenceObject);
+			
+			// Identifier cache
+			if ( !className.equals(ReactomeJavaConstants.Reaction) )
+			{
+				addObjectToIdentifierCache(className, objectCacheByIdentifier, referenceObject);
+			}
+		});
+		logger.info("Built {} caches:"
+				+ "\n\tKeys in cache-by-refdb: {};"
+				+ "\n\tkeys in cache-by-species: {};"
+				+ "\n\tkeys in cache-by-DB_ID: {};"
+				+ "\n\tkeys in cache-by-identifier: {};",className,
+					(cacheByRefDB!=null ? cacheByRefDB.size() : "N/A"),
+					(!className.equals(ReactomeJavaConstants.ReferenceMolecule)
+							&& !className.equals(ReactomeJavaConstants.DatabaseIdentifier)
+							&& !className.equals(ReactomeJavaConstants.ReferenceTherapeutic)
+						? cacheBySpecies.size() : "N/A"),
+					objectCacheByIdentifier.size(),
+					cacheByID.size()
+				);
+	}
+
+	/**
+	 * Adds an object to a cache that caches by identifier.
+	 * @param className - type of object being cached.
+	 * @param objectCacheByIdentifier - the cache to add the object into.
+	 * @param referenceObject - The object that will be added to a cache.
+	 * @throws Error
+	 */
+	private static void addObjectToIdentifierCache(String className, Map<String, List<GKInstance>> objectCacheByIdentifier, GKInstance referenceObject) throws Error
+	{
+		try
+		{
+			if (referenceObject.getSchemClass().isValidAttribute(ReactomeJavaConstants.identifier) && referenceObject.getAttributeValue(ReactomeJavaConstants.identifier) != null)
+			{
+				String identifier = (String) referenceObject.getAttributeValue(ReactomeJavaConstants.identifier);
 				try
 				{
-					refDBID = String.valueOf(((GKInstance) referenceObject.getAttributeValue(ReactomeJavaConstants.referenceDatabase)).getDBID());
-					if (! cacheByRefDB.containsKey(refDBID))
+					if (!objectCacheByIdentifier.containsKey(identifier))
 					{
-						List<GKInstance> byRefDb = Collections.synchronizedList( new LinkedList<GKInstance>() );
-						byRefDb.add(referenceObject);
-						cacheByRefDB.put(refDBID, byRefDb);
+						objectCacheByIdentifier.put(identifier, Collections.synchronizedList(new ArrayList<GKInstance>(Arrays.asList(referenceObject))) );
```

Comment:
I think `Arrays.asList` will return an ArrayList, so it should be safe to remove the wrapping of `new ArrayList<GKInstance>`

Reply:
No, you need to explicitly request something that will be compatible with `Collections.synchronizedList`. If you just use `Arrays.asList` you get this stack trace:
```
java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractList.add(AbstractList.java:153)
	at java.base/java.util.AbstractList.add(AbstractList.java:111)
	at java.base/java.util.Collections$SynchronizedCollection.add(Collections.java:2040)
	at org.reactome.addlinks.db.ReferenceObjectCache.addObjectToIdentifierCache(ReferenceObjectCache.java:166)
...
```

Reply:
Okay, thanks for the explanation and info!